Support keras JSON/HDF5 serialization of main layers

gthb opened the pull request:
Fixes #3101

gthb commented:
Crud, as the test shows, of course load doesn't work in this first-stab implementation, because the layer classes get instantiated with a `dict` instead of a `PretrainedConfig` instance. So needs more work.

gthb commented:
With this change, seven of the 11 `TF*MainLayer` classes pass the test (saving, loading and producing same output for same input after the Keras save/load roundtrip). So I'm just not marking the other four `@keras_serializable` as I haven't gotten it working for those. Specifically:

* TFT5MainLayer does not accept the same model inputs directly, as produced by `self.model_tester.prepare_config_and_inputs_for_common()`, so calling it fails
* `TFXLMMainLayer`, `TFOpenAIGPTMainLayer`, and `TFDistilBertMainLayer` all fail the test (if I add the `@keras_serializable` attribute to them) in the same way: by outputting a tensor of shape `(7, 32)` after the save/load round-trip, which is just the first row of the `(13, 7, 32)` tensor that's output before. I haven't figured out the cause of this.

gthb commented on src/transformers/modeling_tf_openai.py:
```diff
@@ -199,7 +199,7 @@ def call(self, inputs, training=False):
 class TFOpenAIGPTMainLayer(tf.keras.layers.Layer):
     def __init__(self, config, *inputs, **kwargs):
-        super().__init__(config, *inputs, **kwargs)
+        super().__init__(*inputs, **kwargs)
```
Bugfix, independent of this PR and posted separately in #3115, but included here because it got in the way of testing.

gthb commented on src/transformers/modeling_tf_albert.py:
```diff
@@ -478,9 +478,10 @@ def call(self, hidden_states):
 class TFAlbertMainLayer(tf.keras.layers.Layer):
     def __init__(self, config, **kwargs):
-        super().__init__(config, **kwargs)
+        super().__init__(**kwargs)
```
Bugfix, independent of this PR and posted separately in #3115, but included here because it got in the way of testing.

gthb commented on tests/test_modeling_tf_common.py:
```diff
@@ -89,13 +90,49 @@ def test_save_load(self):
+                filepath = os.path.join(tmpdirname, "keras_model.h5")
+                model.save(filepath)
+                model = tf.keras.models.load_model(
+                    filepath, custom_objects={main_layer_class.__name__: main_layer_class}
```
Passing `custom_objects` here _shouldn't_ be necessary because we `return tf.keras.utils.register_keras_serializable()(cls)` in the decorator, but it _is_ necessary with TF 2.1.0 because of https://github.com/tensorflow/tensorflow/issues/36259 whose [fix](https://github.com/tensorflow/tensorflow/pull/37018) only landed two days ago, so it's only fixed in `tf-nightly`.

thomwolf commented on src/transformers/modeling_tf_t5.py:
```diff
@@ -383,14 +383,21 @@ def _prune_heads(self, heads_to_prune):
     ):
-
+        if isinstance(inputs, (tuple, list)):
+            hidden_states = inputs[0]
```
Nice, so normally we have all the inputs possible in the list/dict, i.e. here also attention_mask, encoder_hidden_states, encoder_attention_mask, head_mask see e.g. https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_tf_bert.py#L506-L523

The reason we do that is that it let you use all the features of Keras.fit() like type casting and so on, which are only implemented for the first positional input of the model

thomwolf replied:
Do you think you can add all the inputs here as well like in the example I linked to above.

gthb replied:
Actually the PR shouldn't be touching this class at all, as I backed out of making it Keras-serializable. Will remove.

gthb replied:
Removed.

thomwolf commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -47,6 +47,36 @@ def num_parameters(self, only_trainable: bool = False) -> int:
+    cls._keras_serializable = True
+    if hasattr(tf.keras.utils, "register_keras_serializable"):
+        cls = tf.keras.utils.register_keras_serializable()(cls)
+    return cls
```
I'm not sure about the choice of a wrapper instead of just inheriting from a base `TFMainLayer` like you mentioned in #3101.  I feel like a base class would be more easy to read and understand rather than this complex scheme of multiple wrapping.

gthb replied:
See rationale in https://github.com/huggingface/transformers/pull/3103/commits/0c716ede8c654cd2cdd3ab924f099e5d8277ee68 — I agree that a `TFMainLayer` superclass would be simpler and I tried that first (see first commit in this PR) but soon realized that there's no way to make Keras deserialization work that way, because Keras will pass the JSON-deserialized object (typically a dict) to the initializer — but the initializers of layers in this project expect an instance of `PretrainedConfig` or a subclass of it. A superclass can't affect what arguments are passed to a subclass initializer. But a class decorator can. So that's what I fell back on.

An alternative would be to add (nearly identical) code to _all_ those initializers to handle the dict case as well as the config object case (and document a rule that new main-layer classes should do that also). That seemed messier and less simple than the class decorator approach.

gthb replied:
> rather than this complex scheme of multiple wrapping.

Note, there isn't actually any _multiple_ wrapping — the only thing that gets _wrapped_ is the `__init__` method (to intercept the `config` parameter). The thing that gets returned is the original `cls` object, just with these modifications applied to it:

* wrapping `__init__` (the reason this can't just be a superclass)
* overriding `get_config` to add the `config` attribute
* adding the `_keras_serializable = True` attribute so it can be checked for (currently only used by the test, to decide which layer classes to test)

(Also calls `register_keras_serializable()(cls)` if available, but that just inspects the class and stores references to it, but returns the class itself unmodified. So in principle we could just call `register_keras_serializable()` it and not assign its return value to `cls`, but that would be baking knowledge of its current implementation into this code.)

thomwolf replied:
Ok got it, thanks for the details.

I like your approach!

gthb commented on tests/test_modeling_tf_common.py:
```diff
@@ -89,13 +90,49 @@ def test_save_load(self):
-                out_1 = out_1[~np.isnan(out_1)]
-                out_2 = out_2[~np.isnan(out_2)]
-                max_diff = np.amax(np.abs(out_1 - out_2))
-                self.assertLessEqual(max_diff, 1e-5)
```
> Do you have a very specific reason to remove it or was it more for "cleaning the code"?

I didn't remove it :) — I extracted it to a new method `assert_outputs_same` to share it with the new test I was adding, see line 127.

thomwolf replied:
Ok, I read too quickly, great indeed!

gthb commented on src/transformers/configuration_auto.py:
```diff
@@ -100,6 +101,20 @@ def __init__(self):
         )
 
+    @classmethod
+    def config_class_for_model_class(cls, model_class):
```
Right, that is simpler ... I was trying to change as few places as possible, but that indeed comes at a cost in readability.

gthb replied:
Done.

thomwolf commented:
Ok this looks good to me. 

Do you want to make `make style` and `make quality` to pass the code quality checks (see our [contributor guide](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) if needed), and I guess we can merge it.

gthb commented:
Fixed quality thing, added @functools.wraps (really shouldn't wrap without that as it causes confusing metadata), added a docstring to the `keras_serializable` decorator, and changed to use a name specific to this library rather than the general name `config`, to be clearer where `transformers` is used along with other things in a Keras model. @thomwolf OK with all that?

gthb commented on src/transformers/modeling_tf_t5.py:
```diff
@@ -383,21 +383,14 @@ def _prune_heads(self, heads_to_prune):
     def call(
         self,
-        inputs,
+        hidden_states,
```
No, I don't think so, because the PR is not touching `modeling_tf_t5.py` at all anymore (recall that I reverted the changes in that file in 6fe1cc08743ee8f8b76c622e5a0742413048e857).

I left out `TFT5MainLayer` because, although it fits the _naming pattern_ of other main layers in the project, it doesn't fit the _usage pattern_ (i.e. that the main layer accepts the model's inputs directly). The `test_keras_save_load` test implementation assumes that usage pattern, being based on `self.model_tester.prepare_config_and_inputs_for_common` ... so it fails on the T5 model.

Maybe the `@keras_serializable` decorator would work just fine for TFT5MainLayer, but verifying that would require a test case written more specifically for it, and I didn't sink my teeth into that.

I also left out three other `TF*MainLayer` classes simply because they failed the test and I couldn't quickly see why, see [comment above](https://github.com/huggingface/transformers/pull/3103#issuecomment-594213462)

thomwolf replied:
Ok make sense

thomwolf commented:
Ok this is good to me.
Thanks a lot for the awesome work on this.

Merging @LysandreJik @julien-c

gthb commented:
This landed in 2.6.0 but is missing in the release notes there https://github.com/huggingface/transformers/releases/tag/v2.6.0

LysandreJik commented:
Indeed, that is my fault, I'm sorry for missing it.

I've added it to the v2.7.0 release notes of this morning: https://github.com/huggingface/transformers/releases/tag/v2.7.0

Thanks again for your contribution @gthb !
